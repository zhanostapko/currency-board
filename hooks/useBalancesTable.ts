"use client";
import { useState, useEffect } from "react";
import { BalanceItem } from "@/types/BalanceTableTypes";

export const useBalancesTable = (data: BalanceItem[] = []) => {
  const [columnCount, setColumnCount] = useState(1);
  const [maxColumns, setMaxColumns] = useState(5);
  const [searchQuery, setSearchQuery] = useState("");
  const [balances, setBalances] = useState<BalanceItem[]>(data);
  const [showModal, setShowModal] = useState(false);
  const [currencyToDelete, setCurrencyToDelete] = useState<string>("");

  useEffect(() => {
    const handleResize = () => {
      const width = window.innerWidth;
      if (width < 640) {
        setColumnCount(1);
        setMaxColumns(1);
      } else if (width < 1024) {
        setColumnCount(2);
        setMaxColumns(3);
      } else if (width < 1280) {
        setColumnCount(3);
        setMaxColumns(5);
      } else {
        setColumnCount(4);
        setMaxColumns(5);
      }
    };

    handleResize();
    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  }, []);

  const increment = () =>
    setColumnCount((prev) => Math.min(maxColumns, prev + 1));
  const decrement = () => setColumnCount((prev) => Math.max(1, prev - 1));

  const searchHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(e.target.value);
  };

  const closeModal = () => {
    setShowModal(false);
    setCurrencyToDelete("");
  };

  const showDeleteModal = (currency: string) => {
    setCurrencyToDelete(currency);
    setShowModal(true);
  };

  const handleDelete = (currency: string) => {
    setBalances((prev) => prev.filter((item) => item.currency !== currency));
    closeModal();
  };

  const filteredData = balances.filter((item) =>
    item.currency.includes(searchQuery.toUpperCase())
  );

  return {
    columnCount,
    maxColumns,
    searchQuery,
    balances,
    showModal,
    currencyToDelete,
    setSearchQuery,
    increment,
    decrement,
    showDeleteModal,
    handleDelete,
    closeModal,
    filteredData,
    searchHandler,
  };
};
