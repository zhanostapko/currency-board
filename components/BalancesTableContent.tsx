"use client";
import BalancesTableControls from "./BalancesTableControls";
import BalancesTable from "./BalancesTable";
import DeleteModal from "./DeleteModal";
import { useBalancesTable } from "@/hooks/useBalancesTable";
import { BalanceItem } from "@/types/BalanceTableTypes";

export type BalancesTableContentProps = {
  data?: BalanceItem[];
  error?: string | null;
  title: string;
};

const BalancesTableContent = ({
  data = [],
  error,
  title,
}: BalancesTableContentProps) => {
  const {
    columnCount,
    maxColumns,
    searchQuery,
    filteredData,
    increment,
    decrement,
    showDeleteModal,
    handleDelete,
    closeModal,
    showModal,
    currencyToDelete,
    searchHandler,
  } = useBalancesTable(data);

  return (
    <div className="max-w-7xl mx-auto bg-white p-6 rounded-xl shadow-lg border border-gray-200 w-full text-center">
      <h2 className="text-3xl font-bold text-gray-900 mb-6">{title}</h2>
      <BalancesTableControls
        {...{
          searchQuery,
          onSearchChange: searchHandler,
          columnCount,
          maxColumns,
          increment,
          decrement,
          error,
        }}
      />
      <BalancesTable
        {...{ data: filteredData, columnCount, showDeleteModal, error }}
      />
      {showModal && (
        <DeleteModal
          onConfirm={handleDelete}
          onCancel={closeModal}
          currency={currencyToDelete}
        />
      )}
    </div>
  );
};

export default BalancesTableContent;
