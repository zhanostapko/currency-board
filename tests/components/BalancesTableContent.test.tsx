import { render, screen, fireEvent } from "@testing-library/react";
import BalancesTableContent from "../../components/BalancesTableContent";
import { useBalancesTable } from "@/hooks/useBalancesTable";

jest.mock("../../hooks/useBalancesTable.ts");

describe("BalancesTableContent Component", () => {
  const mockUseBalancesTable = {
    columnCount: 2,
    maxColumns: 5,
    searchQuery: "",
    filteredData: [
      { currency: "USD", amount: 100 },
      { currency: "EUR", amount: 200 },
    ],
    increment: jest.fn(),
    decrement: jest.fn(),
    showDeleteModal: jest.fn(),
    handleDelete: jest.fn(),
    closeModal: jest.fn(),
    showModal: false,
    currencyToDelete: "",
    searchHandler: jest.fn(),
  };

  beforeEach(() => {
    (useBalancesTable as jest.Mock).mockReturnValue(mockUseBalancesTable);
  });

  test("renders correctly and matches snapshot", () => {
    const { asFragment } = render(<BalancesTableContent title="Balances" />);
    expect(screen.getByText("Balances")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("passes correct props to BalancesTableControls", () => {
    render(<BalancesTableContent title="Balances" />);
    expect(screen.getByPlaceholderText("Search")).toBeInTheDocument();
    expect(screen.getByText("+ Increase")).toBeInTheDocument();
    expect(screen.getByText("- Reduce")).toBeInTheDocument();
  });

  test("renders table with filtered data", () => {
    const { asFragment } = render(<BalancesTableContent title="Balances" />);
    expect(screen.getByText("USD")).toBeInTheDocument();
    expect(screen.getByText("100")).toBeInTheDocument();
    expect(screen.getByText("EUR")).toBeInTheDocument();
    expect(screen.getByText("200")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders error message when error is passed", () => {
    const { asFragment } = render(
      <BalancesTableContent title="Balances" error="Test error" />
    );
    expect(
      screen.getByText("❌ Failed to fetch data. Please try again later.")
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows delete modal when showModal is true", () => {
    (useBalancesTable as jest.Mock).mockReturnValue({
      ...mockUseBalancesTable,
      showModal: true,
      currencyToDelete: "USD",
    });

    const { asFragment } = render(<BalancesTableContent title="Balances" />);
    expect(
      screen.getByText(/Are you sure you want to delete/i)
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("calls increment and decrement functions", () => {
    render(<BalancesTableContent title="Balances" />);

    fireEvent.click(screen.getByText("+ Increase"));
    fireEvent.click(screen.getByText("- Reduce"));

    expect(mockUseBalancesTable.increment).toHaveBeenCalledTimes(1);
    expect(mockUseBalancesTable.decrement).toHaveBeenCalledTimes(1);
  });
});
