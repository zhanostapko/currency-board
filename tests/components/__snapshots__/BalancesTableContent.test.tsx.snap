// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalancesTableContent Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-7xl mx-auto bg-white p-6 rounded-xl shadow-lg border border-gray-200 w-full text-center"
  >
    <h2
      class="text-3xl font-bold text-gray-900 mb-6"
    >
      Balances
    </h2>
    <div
      class="flex flex-wrap items-center justify-center gap-4 mb-6"
    >
      <input
        class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
        placeholder="Search"
        value=""
      />
      <div
        class="flex gap-4"
      >
        <button
          class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          + Increase
        </button>
        <button
          class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          - Reduce
        </button>
      </div>
    </div>
     
    <div
      class="w-full overflow-x-auto"
    >
      <table
        class="w-full border-collapse rounded-lg shadow-sm table-fixed"
      >
        <thead>
          <tr
            class="bg-gray-200 text-gray-900 text-lg font-semibold"
          >
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="even:bg-gray-100 hover:bg-gray-200 transition"
          >
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              USD
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              100
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              EUR
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              200
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BalancesTableContent Component renders error message when error is passed 1`] = `
<DocumentFragment>
  <div
    class="max-w-7xl mx-auto bg-white p-6 rounded-xl shadow-lg border border-gray-200 w-full text-center"
  >
    <h2
      class="text-3xl font-bold text-gray-900 mb-6"
    >
      Balances
    </h2>
    <div
      class="flex flex-wrap items-center justify-center gap-4 mb-6"
    >
      <input
        class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
        disabled=""
        placeholder="Search"
        value=""
      />
      <div
        class="flex gap-4"
      >
        <button
          class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
          disabled=""
        >
          + Increase
        </button>
        <button
          class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
          disabled=""
        >
          - Reduce
        </button>
      </div>
    </div>
     
    <div
      class="w-full border border-red-500 bg-red-50 text-red-700 text-center font-medium p-4 rounded-md"
    >
      ‚ùå Failed to fetch data. Please try again later.
    </div>
  </div>
</DocumentFragment>
`;

exports[`BalancesTableContent Component renders table with filtered data 1`] = `
<DocumentFragment>
  <div
    class="max-w-7xl mx-auto bg-white p-6 rounded-xl shadow-lg border border-gray-200 w-full text-center"
  >
    <h2
      class="text-3xl font-bold text-gray-900 mb-6"
    >
      Balances
    </h2>
    <div
      class="flex flex-wrap items-center justify-center gap-4 mb-6"
    >
      <input
        class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
        placeholder="Search"
        value=""
      />
      <div
        class="flex gap-4"
      >
        <button
          class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          + Increase
        </button>
        <button
          class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          - Reduce
        </button>
      </div>
    </div>
     
    <div
      class="w-full overflow-x-auto"
    >
      <table
        class="w-full border-collapse rounded-lg shadow-sm table-fixed"
      >
        <thead>
          <tr
            class="bg-gray-200 text-gray-900 text-lg font-semibold"
          >
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="even:bg-gray-100 hover:bg-gray-200 transition"
          >
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              USD
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              100
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              EUR
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              200
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BalancesTableContent Component shows delete modal when showModal is true 1`] = `
<DocumentFragment>
  <div
    class="max-w-7xl mx-auto bg-white p-6 rounded-xl shadow-lg border border-gray-200 w-full text-center"
  >
    <h2
      class="text-3xl font-bold text-gray-900 mb-6"
    >
      Balances
    </h2>
    <div
      class="flex flex-wrap items-center justify-center gap-4 mb-6"
    >
      <input
        class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
        placeholder="Search"
        value=""
      />
      <div
        class="flex gap-4"
      >
        <button
          class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          + Increase
        </button>
        <button
          class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        >
          - Reduce
        </button>
      </div>
    </div>
     
    <div
      class="w-full overflow-x-auto"
    >
      <table
        class="w-full border-collapse rounded-lg shadow-sm table-fixed"
      >
        <thead>
          <tr
            class="bg-gray-200 text-gray-900 text-lg font-semibold"
          >
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
            <th
              class="px-6 py-4 text-left"
            >
              Currency
            </th>
            <th
              class="px-6 py-4 text-right"
            >
              Balance
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="even:bg-gray-100 hover:bg-gray-200 transition"
          >
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              USD
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              100
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium text-left"
            >
              EUR
            </td>
            <td
              class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
            >
              200
              <button
                class="text-red-500 hover:text-red-700 transition"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  height="14"
                  loading="lazy"
                  src="trashIcon.svg"
                  style="color: transparent;"
                  width="14"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="fixed inset-0 flex items-center justify-center bg-black bg-opacity-50 p-6"
    >
      <div
        class="max-w-lg w-full bg-white p-8 rounded-xl shadow-lg border border-gray-200 text-center"
      >
        <h2
          class="text-2xl font-bold text-gray-900 mb-4"
        >
          Confirm Deletion
        </h2>
        <p
          class="text-gray-600 mb-6"
        >
          Are you sure you want to delete 
          <strong>
            USD
          </strong>
           from the list?
        </p>
        <div
          class="flex justify-center gap-4"
        >
          <button
            class="px-5 py-2.5 text-sm bg-red-500 text-white rounded-md hover:bg-red-600 transition"
          >
            Delete
          </button>
          <button
            class="px-5 py-2.5 text-sm border border-gray-300 bg-white text-gray-700 rounded-md hover:bg-gray-100 transition"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
